Select humanoid tracking example via button

thowell opened the pull request:
Logic to select motion that is tracked in humanoid tracking task. Utilizes `userdata` to store the current motion id in `userdata[0]` and the simulation time in `userdata[1]`. A future PR (after #65 is incorporated) will enabling looping through all tasks.
@hartikainen

hartikainen commented on mjpc/tasks/humanoid/tracking/task.xml:
```diff
@@ -137,14 +138,14 @@
-  <!-- <include file="./keyframes/CMU-CMU-88-88_07_poses.xml" />  <!-\- cartwheel-\-> -->
-  <!-- <include file="./keyframes/CMU-CMU-88-88_08_poses.xml" />  <!-\- crouch and flip backward on hands -\-> -->
-  <!-- <include file="./keyframes/CMU-CMU-88-88_09_poses.xml" />  <!-\- stretch and cartwheel -\-> -->
+  <include file="./keyframes/CMU-CMU-02-02_04_poses.xml" />  simple jump, balance
```
The comment on this line is different from the ones below. Maybe change to `<!-\- simple jump, balance -\->` or remove `<!-\-` and `-\->` from the others

thowell replied:
Standardized the comment style.

hartikainen commented on mjpc/tasks/humanoid/tracking/task.cc:
```diff
@@ -191,10 +243,35 @@ void humanoid::Tracking::Residual(const double *parameters,
+    mju_copy(d->qpos, qpos_home, model->nq);
+    mju_copy(d->qvel, qvel_home, model->nv);
+
+    printf("Motion Switch\n");
```
Maybe print the motion index/name here too.

thowell replied:
Done.

hartikainen commented on mjpc/tasks/humanoid/tracking/task.cc:
```diff
@@ -191,10 +243,35 @@ void humanoid::Tracking::Residual(const double *parameters,
+
+  // indices
+  double current_index = (d->time - d->userdata[1]) * fps + start;  
+  int last_key_index = start + length - 1; // TODO(taylor): check...
```
This seems correct to me (re your `TODO` here).

thowell replied:
Comment removed.

hartikainen commented on mjpc/tasks/humanoid/tracking/task.cc:
```diff
@@ -37,6 +38,55 @@ std::tuple<int, int, double, double> ComputeInterpolationValues(double index,
+    return 510;
+  // Stand
+  } else if (id == 10) {
+    return 1;
```
Is this one missing the actual keyframe?

thowell replied:
Removed.

hartikainen commented on mjpc/tasks/humanoid/tracking/task.cc:
```diff
@@ -37,6 +38,55 @@ std::tuple<int, int, double, double> ComputeInterpolationValues(double index,
+// return length of motion trajectory
+int TrajectoryLength(int id) {
+  // Jump
+  if (id == 0) {
```
Can we comment this also with the the mocap filename ids (e.g. `CMU-CMU-88-88_09_poses.xml`).

thowell replied:
Done.

hartikainen reviewed (commented):
Looks good to me!

hartikainen commented on mjpc/tasks/humanoid/tracking/task.cc:
```diff
@@ -191,10 +243,35 @@ void humanoid::Tracking::Residual(const double *parameters,
+    d->userdata[0] = task->stage; // set motion id
+    d->userdata[1] = d->time;  // reference time
+
+    // set home state
+    double* qpos_home = model->key_qpos + model->nq * start;
+    double* qvel_home = model->key_qvel + model->nv * start;
+
+    mju_copy(d->qpos, qpos_home, model->nq);
+    mju_copy(d->qvel, qvel_home, model->nv);
```
Maybe set with `mj_resetDataKeyframe`?

thowell replied:
Done.

thowell commented:
@hartikainen thanks for you comments. I addressed each one and made most of the changes that you suggested. 
@yuvaltassa can you review and merge?

yuvaltassa commented:
Give me a day or two?, @erez-tom's restructure is nearly ready to submit.

thowell commented:
Updated to merge in latest changes to main. Additionally, updated Quadruped Hill and Quadrotor tasks to use `Task` member `int current_stage` instead of `userdata[0]`.

erez-tom commented on mjpc/tasks/humanoid/tracking/task.xml:
```diff
@@ -1,9 +1,10 @@
   <include file="../../common.xml"/>
   <include file="../humanoid.xml" />
-  <size memory="400K"/>
+  <size memory="400K" nuserdata="2"/> <!-- userdata: [motion id, reference time]-->
```
you probably don't need nuserdata anymore, right?  maybe also in the quadrotor?

thowell replied:
Removed from all task xml files now.
